---
title: Rentrez Tutorial
author: "David winter"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: true
    theme: readable
    figwidth: 3
vignette: >
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEngine{knitr::rmarkdown}
  \usepackage[utf8]{inputenc}
---

```{r, count_recs, echo=FALSE}
count_recs <- function(db, denom) {
    round(as.integer(entrez_db_summary(db)["Count"])/denom, 1)
}
    
```

 
## Introduction: The NCBI, entrez and `rentrez`.

The NCBI stores a _lot_ of data. At the time this document was compiled, there
were `r count_recs("pubmed",1e6)` million papers in [pubmed](http://www.ncbi.nlm.nih.gov/pubmed/), 
including `r count_recs("pmc", 1e6)` million full-text records avaliable in [pubmed-central](http://www.ncbi.nlm.nih.gov/pubmed/). 
[The NCBI Nucleotide Database](http://www.ncbi.nlm.nih.gov/nuccore) (also known as GenBank) has data for `r count_recs("nuccore", 1e6)`
million different sequences, and [dbSNP](http://www.ncbi.nlm.nih.gov/snp/) describes
`r count_recs("snp", 1e6)` million different genetic variants. All of these
records can be cross-referenced with the  `r floor(as.integer(entrez_search(db="taxonomy", term='species[RANK]')$count)/1000)` thousand
species in the [NCBI taxonomy](www.ncbi.nlm.nih.gov/taxonomy) or `r count_recs("omim", 1e3)` thousand diseases
described in [OMIM]('http://www.ncbi.nlm.nih.gov/omim').


The NCBI makes this data avaliable through a [web interface](http://www.ncbi.nlm.nih.gov/), 
an [FTP server](ftp://ftp.ncbi.nlm.nih.gov/) and through a REST API called the
[Entrez Utilities](http://www.ncbi.nlm.nih.gov/books/NBK25500/) (`Eutils` for
short). This package provides functions to wrap that API, allowing users to
gather and combine data from multiple NCBI databases in the confort of a R
session or script.

## Getting started with the rentrez

`rentrez` gives you a few tools that help you to get to know your way around the
NCBI's databases. First, you can use `entrez_dbs()` to find out just what
databases are avaliable:

```{r, dbs}
library(rentrez)
entrez_dbs()
```

You can find out a bit more about each of those database using a set of
functions with names starting `entrez_db_`:

**Functions that help you learn about NCBI databases**

| Function name            | Return                                               |
|--------------------------|------------------------------------------------------|
| `entrez_db_summary()`    | Brief description of what the database is            |
| `entrez_db_searchable()` | Set of search terms that can used with this database |
| `entrez_db_links() `     | Set of databases that might contain linked records   |

For instance, we can find out a little more about the cryptically named
database 'cdd'...

```{r, cdd}
entrez_db_summary("cdd")
```

... or find out which search terms can be used with the Short Read Archive (SRA)
sequencing database:

```{r, sra_eg}
entrez_db_searchable("sra")
```

Just how these 'helper' functions might help you will become clearer once
you've started using the EUtils api, so let's get started.

## Searching databases: `entrez_search`

The function `entrez_search()` lets you search for records in a given NCBI
database.  In the simplest case you just need to provide a database name (`db`) 
and a search term(`term`).If you print the returned object you get a summary of what it contains, an
d the way the NCBI's servers translated your query:

```{r eg_search}
r_search <- entrez_search(db="pubmed", term="R Language")
r_search
```

There are a few things to note here. First, the NCBI's server has worked out
that we meant R as a programming language, and so inlcuded the 
['MeSh' term](http://www.ncbi.nlm.nih.gov/mesh). We'll worry about these and
other special search terms in a second. Second, there are many more 'hits' for this search than there
are unique IDs contained in this object. That's because the optional argument
`retmax`, which controls the maximum number of returned values has a default
value of 20. We can acess those 20 IDs like this:


```{r search_ids}
r_search$ids
```

And get more IDs by increasing the `ret_max` argument. 

```{r searchids_2}
another_r_search <- entrez_search(db="pubmed", term="R Language", retmax=40)
another_r_search
```

If we want to get IDs for all of those thousands of records that match this
search, we can use the NCBI's web history feature, DESCRIBED LATER IN THE
TUTORIAL


### Building search terms

The EUtils api uses a special syntax to build search queries. You can search a
database against a specific term using the format `query[SEARCH TERM]`, and 
combine multiple such seaches using the operators `AND`, `OR` and `NOT`.

For instance, we can find next generation sequence datasets for the (amazing...) ciliate
_Tetrahymena thermophila_ by using the organism ('ORGN') search field:


```{r, Tt}
entrez_search(db="sra", 
              term="Tetrahymena thermophila[ORGN]",
              retmax=0)
```

Find only those records that have been added recently (using the colon to
sepcify a range of values):


```{r, Tt2}
entrez_search(db="sra", 
              term="Tetrahymena thermophila[ORGN] AND 2013:2015[PDAT]",
              retmax=0)
```

Or recently records for either _T. thermophila_ or it's close relative _T.
borealis_ (using parentheses to make ANDs and ORs explicit).


```{r, Tt3}
entrez_search(db="sra", 
              term="(Tetrahymena thermophila[ORGN] OR Tetrahymena borealis[ORGN]) AND 2013:2015[PDAT]",
              retmax=0)
```

The set of search terms avalaliable varies between databases. You can get a list
of avaliable terms or any given data base with `entrez_db_searchable()`

```{r, sra_searchable}
entrez_db_searchable("sra")
```

###Precise queries using MeSH terms

In addition to the seach terms described above, the NCBI allows searches using
[Medical Subject Heading (MeSH)](http://www.ncbi.nlm.nih.gov/mesh) terms. These
terms create a 'controled vocubulary',  and allow users to make very finely
controlled queries of databases. 

For instance, if you were interested in reviewing studies on how a class of 
anti-malarial drugs called Folic Acid Antagonists work against _Plasmodium vivax_ (a
particular species of malarial parasite), you could use this search:

```{r, mesh}
entrez_search(db   = "pubmed",
              term = "(vivax malaria[MeSH]) AND (folic acid antagonists[MeSH])")
```

The complete database of mesh terms is avaliable in EUtils, so you can search
for specific terms with `entrez_search(db='mesh, term =...)` and learn about the
terms you search finds with the tools described below.

### Advanced counting

As you can see above, the  object returned by `entrez_search()` includes the the 
number of records matching a given search. This means you can learn a little
about the composition of, or trends in, the records stored in the NCBI's
databases using only the search utility. For instance, let's track the raise of
the buzzword "connectome" in pubmed, programatically creating  search terms for
the `PDAT` field:

```{r, connectome}
search_year <- function(year, term){
    query <- paste(term, "AND (", year, "[PDAT])")
    entrez_search(db="pubmed", term=query, retmax=0)$count
}

year <- 2008:2014
papers <- sapply(year, search_year, term="Connectome", USE.NAMES=FALSE)

plot(year, as.numeric(papers), type='b', main="The Rise of the Connectome")
```

## Finding cross-references : `entrez_link()`:

###Cross-linked data
###External links

## Getting summary data: `entrez_summary()`

Having foundt the unique IDs for some records via`entrez_search`, you are
probably going to want to learn something about them. The `Eutils` API has two
ways to get information about a record. `entrez_fetch()` returns 'full' records
in varying formats and `entrez_summary()` returns less information but in
relatively simple format. Very often the summary records have the information
you are afer, so `rentrez` provides functions to parse and summarise summary
recods. 


###The summary record

`entrez_summary()` takes a vector of unique IDs for the samples you want to get
summary information from. Let's start by finding out something about the paper
describing [Taxize](https://github.com/ropensci/taxize), using its pubmed ID:


```{r, Summ_1}
taxize_summ <- entrez_summary(db="pubmed", id=24555091)
taxize_summ
```

The object returned by `entrez_summary` behaves like a list, so you can extract
elements using `$`. For instance, we could covert our pubmed ID to another
article ID...

```{r, Summ_2}
taxize_summ$articleids
```
...or see how many times the article has been cited in pubmed-central papers

```{r, Summ_3}
taxize_summ$pmcrefcount
```

###Dealing with many records

If you give `entrez_summary()` a vector with more than one ID you'll get a a
list of summary records back. Let's get those _Plasmodium vivax_ papers we found
in the `entrez_search()` section back, and fetch some summary data on each paper:

```{r, multi_summ}
vivax_search <- entrez_search(db = "pubmed",
                              term = "(vivax malaria[MeSH]) AND (folic acid antagonists[MeSH])")
multi_summs <- entrez_summary(db="pubmed", id=vivax_search$ids)
```

`rentrez` provides a helper function, `extract_from_esummary()` that takes one
or more elements from every summary record in one of these lists. Here is is
working with one...

```{r, multi_summ2}
extract_from_esummary(multi_summs, "fulljournalname")
```
... and several elements:

```{r, multi_summ3}
date_and_cite <- extract_from_esummary(multi_summs, c("pubdate", "pmcrefcount",  "title"))
knitr::kable(head(t(date_and_cite)), row.names=FALSE)
```
##Fetching full records

##Using NCBI's Web History features













